Python/chapter03/p06_animal_shelter/miguelHx.py

```diff
@@ -0,0 +1,222 @@
+"""3.6 - Animal Shelter:
+An animal shelter, which holds only dogs and cats, operates on a strictly
+"first in, first out" basis. People must adopt either the "oldest" (based on arrival time)
+animal of all animals at the shelter, or they can select whether they would prefer
+a dog or a cat (and will receive the oldest animal of that type). They cannot
+select which specific animal they would like. Create the data structures to maintain this system
+and implement operations such as enqueue, dequeueAny, dequeueDog, and
+dequeueCat. You may use the built-in LinkedList data structure.
+"""
+import unittest
+
+from collections import deque
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+
+
+class AnimalTypeEnum(Enum):
+    DOG = 'dog'
+    CAT = 'cat'
+
+
+@dataclass
+class Animal:
+    name: str
+    species: AnimalTypeEnum
+    timestamp: Optional[int] = None
+
+
+class AnimalShelter:
+    """We will have each animal instance
+    maintain a timestamp. This timestamp will
+    be used to determine the globally "oldest"
+    (based on arrival time) animal.
+    """
+    def __init__(self):
+        # FIFO, a queue of animals
+        self.dogs = deque()
+        self.cats = deque()
+        self._species_map = {
+            AnimalTypeEnum.DOG.value: self.dogs,
+            AnimalTypeEnum.CAT.value: self.cats
+        }
+        self._global_timestamp = 0
+        self._size = 0
+    
+    def _enqueue(self, animal: Animal) -> None:
+        # list will grow right [<animal_1>, <animal_2>, ..., <animal_n>]
+        animal.timestamp = self._global_timestamp
+        self._global_timestamp += 1
+        self._species_map[animal.species].append(animal)
+
+    def enqueue(self, *animals: Animal, key: str = '') -> None:
```

Comment:
what's the key for?

Reply:
That was mean't to be taken out, left it there by accident.  It was based on my pre-initial implementation idea where key would be 'dog' or 'cat' and that would be used with some conditional logic.  Something like:
python
```
if key == 'dog':
    self.dogs.append(animal)
if key == 'cat':
    self.cats.append(animal)
```
Soon after I decided it would be better to reference the species information from the animal object
rather than using a redundant 'key' argument.

I was meaning to get rid of it but somehow let it slip.

Removed in 93df0f4509874ba862afd9a6abd1e37834f7c009